overlord/snapstate/component.go

```diff
@@ -156,48 +178,56 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
 			// only needed in the "mount-snap" handler and that is
 			// skipped for local revisions.
 			if err := os.Remove(compSetup.CompPath); err != nil {
-				return nil, err
+				return componentInstallTaskSet{}, err
 			}
 		}
 	}
 
-	// TODO hooks for components
+	compInstalled := snapst.IsComponentInCurrentSeq(compSi.Component)
+
+	if !snapsup.Revert && compInstalled {
+		preRefreshHook := SetupPreRefreshComponentHook(st, snapsup.InstanceName(), compSi.Component.ComponentName)
+		componentTS.beforeLink = append(componentTS.beforeLink, preRefreshHook)
+		addTask(preRefreshHook)
```

Comment:
wondering whether it make sense to make this combo a variant of the addTask local helper? another approach would be to have a flag about being before that is set to false at the right point? but maybe is too implicit?

Reply:
I considered something like this, but I do think it is more readable to just be explicit with how we're collecting the tasks.

Reply:
related: do we have a test that checks that ts vs the other list of tasks on componentTs in aggregate list all the same tasks?

Reply:
No, although that isn't really something I think we could do from the public API? We'd have to re-export `doInstallComponent`, which we could do if you like.

Reply:
we do that for doInstall, so maybe is not a bad idea to have a few tests about doInstallComponent that look at the internal behavior/main behavior of it

Reply:
I've reworked this with a suggestion from @alfonsosanchezbeato so that I think it would be harder for them to get out of sync, in b8a07ddfef10220e3a431c4c160a51abf4cc83c9.